cartographer/mapping_2d/sparse_pose_graph/optimization_problem.cc

```diff
@@ -192,6 +209,27 @@ void OptimizationProblem::Solve(const std::vector<Constraint>& constraints,
             .data());
   }
 
+  // Add penalties for violating odometry
+  for (size_t trajectory_id = 0; trajectory_id != odometry_data_.size();
+       ++trajectory_id) {
+    for (size_t odometry_data_index = 1;
+         odometry_data_index < odometry_data_[trajectory_id].size();
+         ++odometry_data_index) {
+      problem.AddResidualBlock(
+          new ceres::AutoDiffCostFunction<WarpedSpaCostFunction, 3, 3, 3, 2>(
+              new WarpedSpaCostFunction(Constraint::Pose{
+                  odometry_data_[trajectory_id][odometry_data_index - 1]
+                      .pose.inverse() *
+                  odometry_data_[trajectory_id][odometry_data_index]
+                      .pose,
+                  0.3, 30.})),
+          nullptr /* loss function */,
+          C_nodes[trajectory_id][odometry_data_index - 1].data(),
+          C_nodes[trajectory_id][odometry_data_index].data(),
+          C_odometry[trajectory_id].data());
+    }
+  }
+
```

Comment:
Since we now tie together consecutive scans using odometry, the following code should be dropped/changed. As I see it odometry can take the place of
```
node_data_[trajectory_id][node_data_index - 1]
    .initial_point_cloud_pose.inverse() *
node_data_[trajectory_id][node_data_index]
    .initial_point_cloud_pose
```
I'd suggest changing the code to have one loop and either use a new computation of this relative pose using odometry if available, or the old code otherwise.